src/snowflake/cli/plugins/nativeapp/run_processor.py

```diff
@@ -342,27 +308,28 @@ def upgrade_app(
             cc.step(f"Creating new application object {self.app_name} in account.")
 
             if self.app_role != self.package_role:
-                with self.use_role(new_role=self.package_role):
+                with self.use_role(self.package_role):
                     self._execute_query(
-                        f"grant install on application package {self.package_name} to role {self.app_role}"
+                        f"grant install, develop on application package {self.package_name} to role {self.app_role}"
+                    )
+                    self._execute_query(
+                        f"grant usage on schema {self.package_name}.{self.stage_schema} to role {self.app_role}"
+                    )
+                    self._execute_query(
+                        f"grant read on stage {self.stage_fqn} to role {self.app_role}"
```

Comment:
Merging these into one `_execute_queries` is possible, but it removes the ability to test these are being executed with the correct role (we can only assert they were executed at _some_ point).